Mathlib/Tactic/Field.lean

```diff
@@ -0,0 +1,81 @@
+/-
+Copyright (c) 2025 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import Mathlib.Tactic.FieldSimp
+import Mathlib.Tactic.Ring.Basic
+
+
+/-! # A tactic for proving algebraic goals in a field
+
+This file contains the `field` tactic, a finishing tactic which roughly consists of running
+`field_simp; ring1`.
+
+-/
+
+open Lean Meta Qq
+
+namespace Mathlib.Tactic.FieldSimp
+
+open Lean Elab Tactic Lean.Parser.Tactic
+
+/--
+The `field` tactic proves equality goals in (semi-)fields. For example:
+```
+example {x y : ℚ} (hx : x + y ≠ 0) : x / (x + y) + y / (x + y) = 1 := by
+  field
+example {a b : ℝ} (ha : a ≠ 0) : a / (a * b) - 1 / b = 0 := by field
+```
+The scope of the tactic is equality goals which are *universal*, in the sense that they are true in
+any field in which the appropriate denominators don't vanish. (That is, they are consequences purely
+of the field axioms.)
+
+Checking the nonvanishing of the necessary denominators is done using a variety of tricks -- in
+particular this part of the reasoning is non-universal, i.e. can be specific to the field at hand
+(order properties, explicit `≠ 0` hypotheses, `CharZero` if that is known, etc).  The user can also
+provide additional terms to help with the nonzeroness proofs. For example:
+```
+example {K : Type*} [Field K] (hK : ∀ x : K, x ^ 2 + 1 ≠ 0) (x : K) :
+    1 / (x ^ 2 + 1) + x ^ 2 / (x ^ 2 + 1) = 1 := by
+  field [hK]
+```
+
+The `field` tactic is built from the tactics `field_simp` (which clears the denominators) and `ring`
+(which proves equality goals universally true in commutative (semi-)rings). If `field` fails to
+prove your goal, you may still be able to prove your goal by running the `field_simp` and `ring_nf`
+normalizations in some order.  For example, this statement:
+```
+example (a b c d z : ℚ) (H1 : c ≠ 0) (H2 : c * z + d ≠ 0) :
+    (a * z + b) / (c * z + d) - c⁻¹ * (z * c + d)⁻¹
+    = a / c - (c * d + c * c * z)⁻¹ - c⁻¹ * (z * c + d)⁻¹ * (a * d - b * c)
+```
+is not proved by `field` but is proved by `ring_nf at *; field_simp; ring`. -/
+elab (name := field) "field" d:(discharger)? args:(simpArgs)? : tactic => withMainContext do
+  let disch ← parseDischarger d args
+  let s0 ← saveState
+  -- run `field_simp` (only at the top level, not recursively)
+  liftMetaTactic1 (transformAtTarget ((AtomM.run .reducible ∘ reduceEq disch) ·) "field"
+    (failIfUnchanged := False) · default)
+  let s1 ← saveState
+  try
+    -- run `ring1`
```

Comment:
This only runs on the main goal, right? Can you add that to the comment?
(Can we add a test where a positivity obligation that would be provable by ring is not discharged? I can try to come up with one.)

Reply:
I'm not sure I follow here: 

> This only runs on the main goal, right?

That's not specific to this use case -- `ring1` always operates only on the main goal!

> Can we add a test where a positivity obligation that would be provable by ring is not discharged?

`ring` does not prove inequalities/disequalities, so I'm not sure what you mean here about discharging positivity obligations?

Reply:
You're right --- I forgot to include my particular knowledge of ring... sorry for the noise!